File: #1. 모의고사 2회/김민정/problem2_신입사원교육.java
// 풀이 시간: 4분
// 시간 복잡도: O(number*logN + N)
// 코드 설명: 정렬한 후 제일 작은 값 두 개를 뽑아서 처리한다
// 취약한 부분:

import java.util.*;

class Solution {
    public int solution(int[] ability, int number) {
        int answer = 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int a : ability) pq.add(a);

        for (int i = 0; i < number; i++) {
            int a1 = pq.poll();
            int a2 = pq.poll();

            pq.add(a1+a2);
            pq.add(a1+a2);
        }

        for (int i = 0; i < ability.length; i++) {
            answer += pq.poll();
        }

        return answer;
    }
}

File: #1. 모의고사 2회/장효빈/problem2_신입사원_교육 .py
# 풀이 시간: 20분 
# 시간 복잡도: NlogN + 4 * logN * number - 처음 정렬, number만큼 2번 뽑고 2번 넣어야 함
# 코드 설명: 능력치가 가장 낮은 계속해서 선택해야 하므로 pq 자료구조를 선택해야 함.
# 취약한 부분: pq를 빠르게 생각 못함... 그리고 python에서 pq를 안써봐서 문서를 보고 품. 숙련이 필요

import heapq

def solution(ability, number):
    answer = sum(ability)
    heapq.heapify(ability)
    for _ in range(number):
        n1 = heapq.heappop(ability)
        n2 = heapq.heappop(ability)
        answer += n1 + n2
        heapq.heappush(ability, n1 + n2)
        heapq.heappush(ability, n1 + n2)
        
    return answer